File: architecture-skills/design-patterns/examples/observer/observer.swift
// Observer pattern in Swift.
//
// Closures with token-based unsubscription keep it simple. In real
// apps prefer the platform tools: Combine publishers, the Observation
// framework (@Observable), NotificationCenter, or KVO. Stored closures
// strongly capture what they reference — use [weak self] in observers
// that outlive their owner.

import Foundation

final class StockTicker {
    typealias PriceObserver = (String, Double) -> Void

    private var observers: [UUID: PriceObserver] = [:]

    @discardableResult
    func subscribe(_ observer: @escaping PriceObserver) -> UUID {
        let token = UUID()
        observers[token] = observer
        return token
    }

    func unsubscribe(_ token: UUID) {
        observers.removeValue(forKey: token)
    }

    func updatePrice(symbol: String, price: Double) {
        for observer in observers.values {
            observer(symbol, price)
        }
    }
}

final class PriceDisplay {
    let name: String

    init(name: String) {
        self.name = name
    }

    func onPrice(symbol: String, price: Double) {
        print("\(name): \(symbol) is now $\(String(format: "%.2f", price))")
    }
}

func makeAlert(watched: String, threshold: Double) -> StockTicker.PriceObserver {
    return { symbol, price in
        if symbol == watched && price > threshold {
            print("ALERT: \(watched) crossed $\(String(format: "%.2f", threshold)) "
                + "(now $\(String(format: "%.2f", price)))")
        }
    }
}

let ticker = StockTicker()
let display = PriceDisplay(name: "Lobby display")
let displayToken = ticker.subscribe { symbol, price in
    display.onPrice(symbol: symbol, price: price)
}
ticker.subscribe(makeAlert(watched: "ACME", threshold: 100))

ticker.updatePrice(symbol: "ACME", price: 98.5)
ticker.updatePrice(symbol: "ACME", price: 101.25)

ticker.unsubscribe(displayToken)
ticker.updatePrice(symbol: "GLOBEX", price: 45) // only the alert remains

File: architecture-skills/design-patterns/examples/observer/observer.js
// Observer pattern in JavaScript.
//
// Closures are the natural observer in JS; subscribe() returns an
// unsubscribe function (the convention used by RxJS, Redux, etc.).
// The platform equivalents are EventTarget in the browser and
// EventEmitter in Node — reach for those before rolling your own.

class StockTicker {
  #observers = new Set();

  subscribe(observer) {
    this.#observers.add(observer);
    return () => this.#observers.delete(observer);
  }

  updatePrice(symbol, price) {
    for (const observer of this.#observers) {
      observer(symbol, price);
    }
  }
}

class PriceDisplay {
  constructor(name) {
    this.name = name;
  }

  onPrice = (symbol, price) => {
    console.log(`${this.name}: ${symbol} is now $${price.toFixed(2)}`);
  };
}

function makeAlert(watchedSymbol, threshold) {
  return (symbol, price) => {
    if (symbol === watchedSymbol && price > threshold) {
      console.log(
        `ALERT: ${watchedSymbol} crossed $${threshold.toFixed(2)} (now $${price.toFixed(2)})`,
      );
    }
  };
}

const ticker = new StockTicker();
const display = new PriceDisplay("Lobby display");
const unsubscribe = ticker.subscribe(display.onPrice);
ticker.subscribe(makeAlert("ACME", 100));

ticker.updatePrice("ACME", 98.5);
ticker.updatePrice("ACME", 101.25);

unsubscribe();
ticker.updatePrice("GLOBEX", 45); // only the alert remains
